### PR TITLE
stream json from OAI

### PR DESCRIPTION
**Pull request by craterink**
enable streaming w/ meaningful update events from OAI response

Usage:
```python
class SentimentSchema(BaseModel):
    thought: str
    sentiment: str

gpt_json = GPTJSON[SentimentSchema](API_KEY)
messages = [
            GPTMessage(
                role=GPTMessageRole.SYSTEM,
                content=SYSTEM_PROMPT,
            ),
        ]

async for partial_result in gpt_json.stream(messages=messages):
    print(partial_result)
```

JSON supported:
- Only works for a single JSON object with no nested objects and whose keys and values are strings.

example of streaming:
- run `python -m examples.stream_example`

todo:
- [ ] testcases
- [ ] update readme with example usage

future work:
- support streaming for more diverse schemas

### CODE REVIEW

File: gpt_json/types_oai.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+from typing import Dict, List, Optional, Union
```
Comment (craterink) [inline gpt_json/types_oai.py:1]: where best

File: gpt_json/types_streaming.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+from enum import Enum
```
Comment (craterink) [inline gpt_json/types_streaming.py:1]: probs needs a little documentation / maybe refactor

File: examples/stream_example.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+class TutorSchema(BaseModel):
+    student_model: str
+    tutor_response: str
+
+gpt_json = GPTJSON[TutorSchema](API_KEY, model=GPTModelVersion.GPT_4)
+
+async def main():
+    problem = "x^2 + 3 = 12"
+    print(make_blue("\nProblem:"), problem)
```
Comment (piercefreeman) [inline examples/stream_example.py:36]: I like the color-based formatting, but for right now I think let's keep the example directory just logic only and minimize additional code imports. If we want to extend the coverage in the future to something a bit more polished, I'd go in the direction of:
- Importing a `click` dependency
- Rewriting some of the entrypoints to these python files for better discoverability
- Using the `secho(fg="color")` function to do all the logging.

File: examples/stream_example.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    print(make_yellow("\nTeacher's thought process:"))
+    teacher_generator = gpt_json.stream(messages=messages, format_variables={"student_model_key": "student_model", "problem" : problem}) 
+    seen_keys = set()
+    async for partial_teacher in teacher_generator:
+        if partial_teacher.event == StreamEventEnum.OBJECT_CREATED:
+            continue
+
+        if partial_teacher.event == StreamEventEnum.KEY_UPDATED and partial_teacher.updated_key not in seen_keys:
+            print({
```
Comment (piercefreeman) [inline examples/stream_example.py:58]: nit: this dictionary access is a bit hard to follow in the print statement, refactor to another variable outside of the print.

File: gpt_json/gpt.py [hunk 5]
```diff
@@ -117,15 +120,7 @@ async def run(
...
-        # Most requests succeed on the first try but we wrap it locally here in case
-        # there is some temporarily instability with the API. If there are longer periods
-        # of instability, there should be system-wide retries in a daemon.
-        backoff_request_submission = backoff.on_exception(
-            backoff.expo,
-            (RateLimitError, OpenAITimeout, APIConnectionError),
-            max_tries=self.openai_max_retries,
-            on_backoff=handle_backoff
-        )(self.submit_request)
+        backoff_request_submission = await self._backoff_submit_request()
```
Comment (piercefreeman) [inline gpt_json/gpt.py:123]: The triple indirection of `_backoff_submit_request -> backoff -> self.submit_request makes` this path hard to follow. Is this refactor required for some codepath downstream?

File: gpt_json/gpt.py [hunk 6]
```diff
@@ -144,6 +139,58 @@ async def run(
...
+        See `run` for documentation. This method is an async generator wrapper around `run` that streams partial results
+        instead of returning them all at once.
+
+        :return: yields `StreamingObject[SchemaType]`s. 
+        """
+        if self.extract_type != ResponseType.DICTIONARY:
+            raise NotImplementedError("For now, streaming is only supported for dictionary responses.")
+        for field_type in self.schema_model.__annotations__.values():
+            if field_type != str:
+                raise NotImplementedError("For now, streaming is only supported for dictionary responses with string fields.")
```
Comment (piercefreeman) [inline gpt_json/gpt.py:158]: I think list values are reasonable to scope out to another PR.

But before limiting this PR to only string values, I'd like to sketch out what the expected behavior for other value types might look like. Some questions:

- If a number is being sent, presumably the streamer should wait until the whole number is collected before sending it back to the caller?
- If a list is being sent, should it yield after every individual value is completed? If we know that it should be yielding a list of strings, should we send those incrementally? (Probably yes) But lists of numbers should probably be queued at least until each value is completed. So there's some compositionally to the expected yielding based on the type value that is currently being processed.

Concretely, let's write up a few simple test cases with different schemas and different expected streaming patterns? We can go to the implementation from there depending on complexity. Let's discuss further in the original issue.

File: gpt_json/types_oai.py (hunk 1 above)
Comment (piercefreeman) [reply gpt_json/types_oai.py:1]: The current project target is > 3.10, so native typing should be supported. Instead of `List/Dict` you can just use `list/dict`, and Optional can just be `Value | None`.

File: gpt_json/types_oai.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+class ChatCompletionDelta:
+    def __init__(
+        self,
+        content: str | None,
+        role: str | None
+    ):
+        self.content = content
+        self.role = role
+
+class ChatCompletionChunkChoice:
```
Comment (piercefreeman) [inline gpt_json/types_oai.py:13]: `ChatCompletionDelta`, `ChatCompletionChunkChoice` and `ChatCompletionChunk` have pretty simple implementations here. They should either be `@dataclass` instances or `pydantic.BaseModel` instances.

File: gpt_json/types_oai.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        object: str,
+    ):
+        self.choices = choices
+        self.created = created
+        self.id = id
+        self.model = model
+        self.object = object
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Union[List[Dict[str, Union[Dict[str, str], None, int]]], int, str]]) -> "ChatCompletionChunk":
```
Comment (piercefreeman) [inline gpt_json/types_oai.py:40]: Given this function's complexity this seems better as a BaseModel.

It looks like you're copying the fields 1:1 from the JSON payload coming back from OpenAI, so you might just be able to make use of BaseModel's default from_dict implementation to handle the verification of parameters and nested types for you?

File: gpt_json/parsers.py [hunk 9]
```diff
@@ -35,3 +37,76 @@ def find_json_response(full_response, extract_type):
...
+
+
+def _is_valid_json(substring: str) -> bool:
+    try:
+        json.loads(substring)
+        return True
+    except json.decoder.JSONDecodeError:
+        return False
+
+def _fix_broken_json(substring: str) -> tuple[str, str]:
```
Comment (piercefreeman) [inline gpt_json/parsers.py:49]: This logic duplicates some portion of what is implemented in `gpt_json.transformations.fix_truncated_json`. Instead of reinventing the wheel here, I'd suggest adding additional coverage and functionality to the parser within `build_stack`.

Specifically - based on a brief skim here - I think you'll want to:
- Detect if we are before an unescaped colon or after, which indicates that we are adding a key
- After a colon but before a comma or closing bracket `}` which indicates that the value is still pending. This will let you wait until full integer fields have been yielded.
- Aside from that you write a sibling function to `fix_truncated_json` that will also inject valid JSON values like `null` if no value is yet supplied

File: gpt_json/types_streaming.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+from pydantic import BaseModel
+
+SchemaType = TypeVar("SchemaType", bound=BaseModel)
+
+class StreamEventEnum(Enum):
+    OBJECT_CREATED = "OBJECT_CREATED"
+    KEY_UPDATED = "KEY_UPDATED"
+    KEY_COMPLETED = "KEY_COMPLETED"
+    OTHER = "OTHER"
```
Comment (piercefreeman) [inline gpt_json/types_streaming.py:12]: This enum value is a bit vague. Seems like it's currently used before we have started streaming values? Can you define what it's intended to be more precisely?

File: gpt_json/types_streaming.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+class StreamingObject(Generic[SchemaType]):
+  schema_model: Type[SchemaType] = None
+
+  updated_key: str | None
+  value_change: str | None
+  event: StreamEventEnum
+  partial_obj: SchemaType
+
+  def __init__(self, obj_data: dict, prev_partial: 'StreamingObject[SchemaType]', proposed_event: StreamEventEnum, value_change: str | None = None) -> None:
```
Comment (piercefreeman) [inline gpt_json/types_streaming.py:23]: nit: specify expected types for the data obj

File: gpt_json/types_streaming.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+     
+     self.event = proposed_event
+     if proposed_event == StreamEventEnum.KEY_UPDATED and self.updated_key is None:
+        # when there updated key is None, we haven't fully streamed the object's first key yet 
+        self.event = StreamEventEnum.OTHER
+     
+     # TODO: this is hacky. ideally we want pydantic to implement Partial[SchemaType]
+     # https://github.com/pydantic/pydantic/issues/1673
+     # my fix is to create the schema object with all string values for now
+     cleaned_obj_data = {field: "" for field in self.schema_model.__fields__.keys()}
```
Comment (piercefreeman) [inline gpt_json/types_streaming.py:38]: Any reason you can't just do the following to replace the dict comprehensions in 38 and 39:

```
cleaned_obj_data = {field: obj_data.get(field, "") for field in self.schema_model.__fields__.keys()}
```

File: gpt_json/gpt.py [hunk 13]
```diff
@@ -144,6 +139,58 @@ async def run(
...
+                # Ignore assistant role message
+                continue
+            if response.choices[0].finish_reason is not None:
+                # Ignore finish message 
+                continue
+            
+            cumulative_response += response.choices[0].delta.content
+            
+            partial_data, event = parse_streamed_json(cumulative_response)
+            partial_response = self.streaming_type(partial_data, prev_partial, event)
```
Comment (piercefreeman) [inline gpt_json/gpt.py:188]: Why are you adding this as an instance method versus just dynamically creating the `StreamingObject` at the start of this function? Is there something with the typehinting that was not validating if you try to create inline?

**Review by piercefreeman — COMMENTED** [review]
Thanks for the PR @craterink! Leaving some more detailed comments inline.

File: gpt_json/types_oai.py (hunk 1 above)
Comment (craterink) [reply gpt_json/types_oai.py:1]: ah I meant the entire contents of the file

File: gpt_json/types_oai.py (hunk 8 above)
Comment (craterink) [reply gpt_json/types_oai.py:40]: omg that's so much better lol

File: gpt_json/parsers.py (hunk 9 above)
Comment (craterink) [reply gpt_json/parsers.py:49]: ah didn't notice the overlap, this is sick. your `build_stack` is way more general too lol

File: gpt_json/gpt.py (hunk 6 above)
Comment (craterink) [reply gpt_json/gpt.py:158]: hm maybe for a list, `updated_key = 0, 1, 2, etc`. for arbitrarily nested objects, let's a list of schema objects, `updated_key = (0, "schema_key_1"), (0, "schema_key_2"), (1, "schema_key_1"), etc"`

Comment (craterink) [reply gpt_json/gpt.py:165]: took a stab at testcases for singly-nested response streaming with arbitrary leaf types (see `gpt_json/tests/test_streaming.json`) -- I still need to test `gpt_json.stream` by mocking OAI completionchunks etc, but I wanted to check in to see what you think of the streaming testcases first

**Review by craterink — COMMENTED** [review]
I think it's ready for another review!

File: gpt_json/streaming.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+  schema_model: Type[SchemaType] = None
+     
+  def __class_getitem__(cls, item):
+    new_cls = super().__class_getitem__(item)
+    new_cls.schema_model = item
+    return new_cls
+
+
+def prepare_streaming_object(schema_model: SchemaType, curr_partial_raw: dict[str, Any], prev_partial: 'StreamingObject[SchemaType]', proposed_event: StreamEventEnum) -> StreamingObject[SchemaType]:
```
Comment (piercefreeman) [inline gpt_json/streaming.py:34]: For the prev_partial typehint, is there a reason it's in a quoted string here? The return type looks the same and seems to validate.

Comment (piercefreeman) [inline gpt_json/streaming.py:34]: Can you add a docstring with the definition of these function parameters?

File: gpt_json/streaming.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+   # compute which key was most recently updated 
+   raw_recent_key = list(curr_partial_raw.keys())[-1] if curr_partial_raw else None
+   updated_key = raw_recent_key if raw_recent_key in schema_model.__fields__ else None
```
Comment (piercefreeman) [inline gpt_json/streaming.py:37]: When would the raw recent not be in the model schema? If GPT is hallucinating results?

File: gpt_json/streaming.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+   event = proposed_event
+   if proposed_event == StreamEventEnum.KEY_UPDATED and updated_key is None:
+      # when the updated key is None, we haven't fully streamed the object's first key yet 
+      event = StreamEventEnum.AWAITING_FIRST_KEY
+   
+   # TODO: this is hacky. ideally we want pydantic to implement Partial[SchemaType]
+   # https://github.com/pydantic/pydantic/issues/1673
+   # my fix is to create the schema object with all string values for now
+   cleaned_obj_data = {field: "" for field, typ in schema_model.__fields__.items()}
+   cleaned_obj_data.update({k:v for k,v in curr_partial_raw.items() if v is not None})
```
Comment (piercefreeman) [inline gpt_json/streaming.py:48]: Good workaround. Mind refactoring into a separate small utility function so we can easily hot-swap it once pydantic closes 1673?

File: gpt_json/streaming.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+   partial_obj = schema_model(**cleaned_obj_data)
+
+   # compute value update if relevant
+   value_change = None
+   if event in [StreamEventEnum.KEY_UPDATED, StreamEventEnum.KEY_COMPLETED]:
+      prev_value = prev_partial.partial_obj.dict()[updated_key]
+      curr_value = partial_obj.dict()[updated_key]
+      if type(prev_value) == str and type(curr_value) == str:
```
Comment (piercefreeman) [inline gpt_json/streaming.py:56]: Doesn't the current method only support string types? When will this value be non-string?

Also, can you use `isinstance(X,Y)` here instead of checking for exact type equality?

File: gpt_json/tests/test_streaming.py [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+    return [enc.decode([tok]) for tok in enc.encode(text)]
+
+def _tuple_merge(t1, t2):
+    t1 = (t1,) if not isinstance(t1, tuple) else t1
+    t2 = (t2,) if not isinstance(t2, tuple) else t2
+
+    return t1 + t2
+
+def _get_expected_stream_partial_objs(full_object):
+    """This function implements the semantics of the behavior we expect from GPTJSON.stream().
```
Comment (piercefreeman) [inline gpt_json/tests/test_streaming.py:26]: This function's getting a bit long, mind refactoring into a separate class:

I'm thinking something like:

```
class ExpectedPartialObjectStreamHarness:
    def __call__(self, full_obj: Any):
        if isinstance(full_obj, list):
            yield from self.handle_list(full_obj)
        elif isinstance(full_obj, dict):
            yield from self.handle_dict(full_obj)

    def handle_list(self, full_obj):
        value_iterators = [self(v) for v in full_object]
        ...
```

I imagine this might also provide better typehinting support.

File: gpt_json/tests/test_streaming.py [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+import tiktoken
+from pydantic import BaseModel
+
+from gpt_json.gpt import GPTJSON
+from gpt_json.models import GPTMessage, GPTMessageRole, GPTModelVersion
+from gpt_json.streaming import StreamEventEnum, StreamingObject
+
+enc = tiktoken.get_encoding("cl100k_base")
+
+def _tokenize(text):
```
Comment (piercefreeman) [inline gpt_json/tests/test_streaming.py:16]: Consider refactoring these into a separate file, like `streaming_helpers.py` to keep a clean line between what streaming code you're testing and what helps you test it. I see you also have a few unit tests for these functions by themselves (which is a great idea) - you could put these into something like `test_streaming_helpers.py` to keep the logic/test separation of concerns.

File: gpt_json/tests/test_streaming.py [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+
+class ExampleInnerSchema(BaseModel):
+    very_nested: int
+    inner_list: list[float]
+class ExampleMultiNestedSchema(BaseModel):
+    is_this_overengineered: bool
+    inner_object: ExampleInnerSchema
+
+EXAMPLE_MULTI_NESTED = {
+    'is_this_overengineered' : True,
```
Comment (piercefreeman) [inline gpt_json/tests/test_streaming.py:146]: Big fan of this key. 😂

File: gpt_json/tests/test_streaming.py [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+
+    # Define mock response
+    mocked_oai_raw_responses = _mock_oai_streaming_chunks(full_object)
+    async def async_list_to_generator(my_list):
+        for item in my_list:
+            yield item
+    mock_response = async_list_to_generator(mocked_oai_raw_responses)
+
+    # Create the mock
+    with patch.object(openai.ChatCompletion, "acreate", return_value=mock_response) as mock_acreate:
```
Comment (piercefreeman) [inline gpt_json/tests/test_streaming.py:230]: I rewrote some of these patches as part of a PR today; since Python got to 3.10 or so they support these async mocks natively:
https://github.com/piercefreeman/gpt-json/pull/11/files#diff-332ec0419d31fe7c72492cf327fc8f968e7122b763badaccba0821c8118cc09dR153

File: gpt_json/tests/test_transformations.py [hunk 22]
```diff
@@ -17,37 +17,50 @@
...
-        ('[{"key1": [123]', [{"key1": [123]}]),
-        ('{"key1": [\n"abc",\n "def', {"key1": ["abc", "def"]}),
+        ('[{"key1": [123]', [{"key1": [123]}], JsonFixEnum.UNCLOSED_OBJECT),
+        ('{"key1": [\n"abc",\n "def', {"key1": ["abc", "def"]}, JsonFixEnum.UNCLOSED_VALUE),
+        # streaming-specific examples
+        ('[\n', [], JsonFixEnum.UNCLOSED_OBJECT),
+        ('[\n', [], JsonFixEnum.UNCLOSED_OBJECT),
+        ('{\n', {}, JsonFixEnum.UNCLOSED_OBJECT),
+        ('[{"broken_key', [{"broken_key": None}], JsonFixEnum.UNCLOSED_KEY),
+        ('[{"key_with_no_value":', [{"key_with_no_value": None}], JsonFixEnum.MISSING_VALUE),
```
Comment (piercefreeman) [inline gpt_json/tests/test_transformations.py:46]: Nice test case expansion! And good move with returning an explanation payload, I think this will come in handy in the future.

File: gpt_json/transformations.py [hunk 23]
```diff
@@ -28,34 +32,58 @@ def is_truncated(json_str):
...
-    # If we still have nested items remaining in our stack,
-    # unwind it into the fixed string
-    if stack:
-        # Unwind the stack by filling it with the closing character
-        # of the current nested level
-        close_stack = ["]" if char == "[" else "}" for char in stack]
-        fixed_str += ''.join(close_stack[::-1])
-
-    return fixed_str, True
+        proposed_fixed_strs = [fixed_str + '"', fixed_str + '": null']
```
Comment (piercefreeman) [inline gpt_json/transformations.py:58]: This manual string manipulation for null values makes me a bit nervous about future extensibility.

Would the following algorithm work to leverage the stack approach:

1. Within `build_stack` keep track of whether we are outside of a quoted string, and before a colon. This must be the key.
2. After the colon and before an unescaped comma or end bracket, we must be in the value (whether quoted, nested, or numerical).
3. If we are have seen the colon but not yet issued a value, we should just append a null character.

**Review by piercefreeman — COMMENTED** [review]
Nice refactor! Left a lot of minor comments. The main chunk remaining as I see it is whether there is a cleaner algorithm for the truncation transformation that sets the stage for the next chapter of extensions. Overall aligned that this is a good V0 scope that can be extended in the future.

File: gpt_json/streaming.py (hunk 15 above)
Comment (craterink) [reply gpt_json/streaming.py:37]: or, if the key has been partially streamed. e.g.:

{
"student_model" : "blah"
}

`student` is a separate token, and so at some point we see `curr_partial_raw` like this:

{
"student": ""
}

File: gpt_json/tests/test_streaming.py (hunk 18 above)
Comment (craterink) [reply gpt_json/tests/test_streaming.py:26]: oh hell ya, nice

**Review by piercefreeman — APPROVED** [review]
LGTM. Thanks Cooper!
